<a href="https://colab.research.google.com/github/AlessandroVol23/ieee_cis_fraud_detection_kaggle/blob/master/notebooks/0_5_AV_TF_Baseline.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Baseline with Tensorflow

- In this notebook I'll build a baseline with tensorflow keras.
- For this kind of challenge it is very unlikely that deep learning will perform better then gradient boosting but I want to get some experience with tensorflow and it would make sense for stacking different models

In [0]:
import os

import numpy as np
import pandas as pd
from sklearn import preprocessing
import xgboost as xgb

import tensorflow as tf

## Functions

In [0]:
# Function to reduce the DF size
def reduce_mem_usage(df, verbose=True):
    numerics = ['int16', 'int32', 'int64', 'float16', 'float32', 'float64']
    start_mem = df.memory_usage().sum() / 1024**2    
    for col in df.columns:
        col_type = df[col].dtypes
        if col_type in numerics:
            c_min = df[col].min()
            c_max = df[col].max()
            if str(col_type)[:3] == 'int':
                if c_min > np.iinfo(np.int8).min and c_max < np.iinfo(np.int8).max:
                    df[col] = df[col].astype(np.int8)
                elif c_min > np.iinfo(np.int16).min and c_max < np.iinfo(np.int16).max:
                    df[col] = df[col].astype(np.int16)
                elif c_min > np.iinfo(np.int32).min and c_max < np.iinfo(np.int32).max:
                    df[col] = df[col].astype(np.int32)
                elif c_min > np.iinfo(np.int64).min and c_max < np.iinfo(np.int64).max:
                    df[col] = df[col].astype(np.int64)  
            else:
                if c_min > np.finfo(np.float16).min and c_max < np.finfo(np.float16).max:
                    df[col] = df[col].astype(np.float16)
                elif c_min > np.finfo(np.float32).min and c_max < np.finfo(np.float32).max:
                    df[col] = df[col].astype(np.float32)
                else:
                    df[col] = df[col].astype(np.float64)    
    end_mem = df.memory_usage().sum() / 1024**2
    if verbose: print('Mem. usage decreased to {:5.2f} Mb ({:.1f}% reduction)'.format(end_mem, 100 * (start_mem - end_mem) / start_mem))
    return df

## colab kaggle init

In [3]:
import os.path

if not os.path.exists('kaggle.json'):
  print("Kaggle Folder doesn't exist yet")
  from google.colab import files
  print("Please click on button an upload your kaggle.json api file")
  files.upload()
  
  !mkdir -p ~/.kaggle
  !cp kaggle.json ~/.kaggle/
  !chmod 600 ~/.kaggle/kaggle.json
  !ls ~/.kaggle
  
  !pip install -q kaggle
  !pip install -q kaggle-cli
  
  !kaggle competitions download -c ieee-fraud-detection
  
  !unzip \*.zip
  
  from IPython.display import clear_output
  clear_output()
  
  print("DONE!")
  
else:
  print("Data already exists")

DONE!


## Data

In [0]:
df_train_ident = pd.read_csv('train_identity.csv', index_col='TransactionID')
df_test_ident = pd.read_csv('test_identity.csv', index_col='TransactionID')

df_train_trans = pd.read_csv('train_transaction.csv', index_col='TransactionID')
df_test_trans = pd.read_csv('test_transaction.csv', index_col='TransactionID')

df_sample_submission = pd.read_csv('sample_submission.csv', index_col='TransactionID')

## Preprocessing

In [5]:
df_train = df_train_trans.merge(df_train_ident, how='left', left_index=True, right_index=True)
df_test = df_test_trans.merge(df_test_ident, how='left', left_index=True, right_index=True)


print(df_train.shape)
print(df_test.shape)

(590540, 433)
(506691, 432)


In [6]:
df_train = reduce_mem_usage(df_train)
df_test = reduce_mem_usage(df_test)

Mem. usage decreased to 668.22 Mb (66.2% reduction)
Mem. usage decreased to 583.43 Mb (65.6% reduction)


In [0]:
emails = {'gmail': 'google', 'att.net': 'att', 'twc.com': 'spectrum', 
          'scranton.edu': 'other', 'optonline.net': 'other', 'hotmail.co.uk': 'microsoft',
          'comcast.net': 'other', 'yahoo.com.mx': 'yahoo', 'yahoo.fr': 'yahoo',
          'yahoo.es': 'yahoo', 'charter.net': 'spectrum', 'live.com': 'microsoft', 
          'aim.com': 'aol', 'hotmail.de': 'microsoft', 'centurylink.net': 'centurylink',
          'gmail.com': 'google', 'me.com': 'apple', 'earthlink.net': 'other', 'gmx.de': 'other',
          'web.de': 'other', 'cfl.rr.com': 'other', 'hotmail.com': 'microsoft', 
          'protonmail.com': 'other', 'hotmail.fr': 'microsoft', 'windstream.net': 'other', 
          'outlook.es': 'microsoft', 'yahoo.co.jp': 'yahoo', 'yahoo.de': 'yahoo',
          'servicios-ta.com': 'other', 'netzero.net': 'other', 'suddenlink.net': 'other',
          'roadrunner.com': 'other', 'sc.rr.com': 'other', 'live.fr': 'microsoft',
          'verizon.net': 'yahoo', 'msn.com': 'microsoft', 'q.com': 'centurylink', 
          'prodigy.net.mx': 'att', 'frontier.com': 'yahoo', 'anonymous.com': 'other', 
          'rocketmail.com': 'yahoo', 'sbcglobal.net': 'att', 'frontiernet.net': 'yahoo', 
          'ymail.com': 'yahoo', 'outlook.com': 'microsoft', 'mail.com': 'other', 
          'bellsouth.net': 'other', 'embarqmail.com': 'centurylink', 'cableone.net': 'other', 
          'hotmail.es': 'microsoft', 'mac.com': 'apple', 'yahoo.co.uk': 'yahoo', 'netzero.com': 'other', 
          'yahoo.com': 'yahoo', 'live.com.mx': 'microsoft', 'ptd.net': 'other', 'cox.net': 'other',
          'aol.com': 'aol', 'juno.com': 'other', 'icloud.com': 'apple'}

us_emails = ['gmail', 'net', 'edu']

# https://www.kaggle.com/c/ieee-fraud-detection/discussion/100499#latest-579654
for c in ['P_emaildomain', 'R_emaildomain']:
    df_train[c + '_bin'] = df_train[c].map(emails)
    df_test[c + '_bin'] = df_test[c].map(emails)
    
    df_train[c + '_suffix'] = df_train[c].map(lambda x: str(x).split('.')[-1])
    df_test[c + '_suffix'] = df_test[c].map(lambda x: str(x).split('.')[-1])
    
    df_train[c + '_suffix'] = df_train[c + '_suffix'].map(lambda x: x if str(x) not in us_emails else 'us')
    df_test[c + '_suffix'] = df_test[c + '_suffix'].map(lambda x: x if str(x) not in us_emails else 'us')

In [8]:
df_train.head(10)

isFraud  TransactionDT  ...  R_emaildomain_bin R_emaildomain_suffix
TransactionID                          ...                                        
2987000              0          86400  ...                NaN                  nan
2987001              0          86401  ...                NaN                  nan
2987002              0          86469  ...                NaN                  nan
2987003              0          86499  ...                NaN                  nan
2987004              0          86506  ...                NaN                  nan
2987005              0          86510  ...                NaN                  nan
2987006              0          86522  ...                NaN                  nan
2987007              0          86529  ...                NaN                  nan
2987008              0          86535  ...                NaN                  nan
2987009              0          86536  ...                NaN                  nan

[10 rows x 437 columns]

In [0]:
# Encode all categorical features
for f in df_train.drop('isFraud', axis=1).columns:
    if df_train[f].dtype=='object' or df_test[f].dtype=='object': 
        lbl = preprocessing.LabelEncoder()
        lbl.fit(list(df_train[f].values) + list(df_test[f].values))
        df_train[f] = lbl.transform(list(df_train[f].values))
        df_test[f] = lbl.transform(list(df_test[f].values))   

In [0]:
# Some feature engineering

df_train['Trans_min_mean'] = df_train['TransactionAmt'] - df_train['TransactionAmt'].mean()
df_train['Trans_min_std'] = df_train['Trans_min_mean'] / df_train['TransactionAmt'].std()
df_test['Trans_min_mean'] = df_test['TransactionAmt'] - df_test['TransactionAmt'].mean()
df_test['Trans_min_std'] = df_test['Trans_min_mean'] / df_test['TransactionAmt'].std()

df_train['TransactionAmt_to_mean_card1'] = df_train['TransactionAmt'] / df_train.groupby(['card1'])['TransactionAmt'].transform('mean')
df_train['TransactionAmt_to_mean_card4'] = df_train['TransactionAmt'] / df_train.groupby(['card4'])['TransactionAmt'].transform('mean')
df_train['TransactionAmt_to_std_card1'] = df_train['TransactionAmt'] / df_train.groupby(['card1'])['TransactionAmt'].transform('std')
df_train['TransactionAmt_to_std_card4'] = df_train['TransactionAmt'] / df_train.groupby(['card4'])['TransactionAmt'].transform('std')

df_test['TransactionAmt_to_mean_card1'] = df_test['TransactionAmt'] / df_test.groupby(['card1'])['TransactionAmt'].transform('mean')
df_test['TransactionAmt_to_mean_card4'] = df_test['TransactionAmt'] / df_test.groupby(['card4'])['TransactionAmt'].transform('mean')
df_test['TransactionAmt_to_std_card1'] = df_test['TransactionAmt'] / df_test.groupby(['card1'])['TransactionAmt'].transform('std')
df_test['TransactionAmt_to_std_card4'] = df_test['TransactionAmt'] / df_test.groupby(['card4'])['TransactionAmt'].transform('std')

In [11]:
df_train.head()

isFraud  ...  TransactionAmt_to_std_card4
TransactionID           ...                             
2987000              0  ...                     0.000000
2987001              0  ...                     0.114258
2987002              0  ...                     0.258545
2987003              0  ...                     0.196899
2987004              0  ...                     0.196899

[5 rows x 443 columns]

In [0]:
df_train['TransactionAmt'] = np.log(df_train['TransactionAmt'])
df_test['TransactionAmt'] = np.log(df_test['TransactionAmt'])

## Model

In [13]:
df_train.head()

isFraud  ...  TransactionAmt_to_std_card4
TransactionID           ...                             
2987000              0  ...                     0.000000
2987001              0  ...                     0.114258
2987002              0  ...                     0.258545
2987003              0  ...                     0.196899
2987004              0  ...                     0.196899

[5 rows x 443 columns]

In [14]:
df_test.head()

TransactionDT  ...  TransactionAmt_to_std_card4
TransactionID                 ...                             
3663549             18403224  ...                     0.129883
3663550             18403263  ...                     0.199219
3663551             18403310  ...                     0.695312
3663552             18403310  ...                     1.158203
3663553             18403317  ...                     0.277344

[5 rows x 442 columns]

In [15]:
print(df_train.shape)
print(df_test.shape)

(590540, 443)
(506691, 442)


In [0]:
# Copy label
y_train = df_train['isFraud'].copy()

# Delete label
X_train = df_train.drop('isFraud', axis=1)

# Create X
X_test = df_test.copy()

del df_train, df_test, df_train_ident, df_test_ident, df_train_trans, df_test_trans

In [17]:
X_train.head()

TransactionDT  ...  TransactionAmt_to_std_card4
TransactionID                 ...                             
2987000                86400  ...                     0.000000
2987001                86401  ...                     0.114258
2987002                86469  ...                     0.258545
2987003                86499  ...                     0.196899
2987004                86506  ...                     0.196899

[5 rows x 442 columns]

In [0]:
sample_X = X_train.copy()

In [0]:
sample_X.fillna(-1, inplace=True)

In [0]:
sample_X[:] = np.nan_to_num(sample_X)

In [22]:
sample_X.shape

(590540, 442)

In [23]:
_END_ROWS = 440
x = sample_X.values[:, :_END_ROWS]
y = y_train.values[:]

print("Null values in x: ", np.isnan(x).sum())
print("Null values in y: ", np.isnan(y).sum())

print("Shape x: ", x.shape)
print("Shape y: ", y.shape)

Null values in x:  0
Null values in y:  0
Shape x:  (590540, 440)
Shape y:  (590540,)


## Normalise Data

In [0]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
x = scaler.fit_transform(x)


## Test Train Split

In [0]:
from sklearn.model_selection import train_test_split
X_tr, X_val, y_tr, y_val = train_test_split(
    x, y, test_size=0.33, random_state=42)

## Tensorflow Model

In [0]:
import datetime
logdir = os.path.join("logs", datetime.datetime.now().strftime("%Y%m%d-%H%M%S"))
tensorboard_callback = tf.keras.callbacks.TensorBoard(logdir, histogram_freq=1)

# Roc-AUC Callback
from sklearn.metrics import roc_auc_score
from keras.callbacks import Callback
class roc_callback(Callback):
    def __init__(self,training_data, validation_data):
        self.x = training_data[0]
        self.y = training_data[1]
        self.x_val = validation_data[0]
        self.y_val = validation_data[1]


    def on_train_begin(self, logs={}):
        return

    def on_train_end(self, logs={}):
        return

    def on_epoch_begin(self, epoch, logs={}):
        return

    def on_epoch_end(self, epoch, logs={}):
        y_pred = self.model.predict(self.x)
        roc = roc_auc_score(self.y, y_pred)
        y_pred_val = self.model.predict(self.x_val)
        roc_val = roc_auc_score(self.y_val, y_pred_val)
        print('\rroc-auc: %s - roc-auc_val: %s' % (str(round(roc,4)),str(round(roc_val,4))),end=100*' '+'\n')
        print('\rroc-auc: %s - roc-auc_val: %s' % (str(round(roc,4)),str(round(roc_val,4))),end=100*' '+'\n')
        # print("\n" + "ROC-AUC Score is: {}".format(str(round(roc, 4))) + "\n")
        return

    def on_batch_begin(self, batch, logs={}):
        return

    def on_batch_end(self, batch, logs={}):
        return
   

In [0]:
from tensorflow import keras
model = tf.keras.Sequential(
    [
        keras.layers.Dense(units=512, input_shape=[_END_ROWS], activation='sigmoid'),
        keras.layers.BatchNormalization(),
        keras.layers.Dense(units=256, activation='sigmoid'),
        keras.layers.BatchNormalization(),
        keras.layers.Dense(units=128, activation='sigmoid'),
        keras.layers.Dense(units=1, activation='sigmoid')
    ]
)

from tensorflow.keras.optimizers import RMSprop

model.compile(loss='binary_crossentropy',
              optimizer=RMSprop(lr=0.001),
              metrics=['acc'])

In [34]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_4 (Dense)              (None, 512)               225792    
_________________________________________________________________
batch_normalization_2 (Batch (None, 512)               2048      
_________________________________________________________________
dense_5 (Dense)              (None, 256)               131328    
_________________________________________________________________
batch_normalization_3 (Batch (None, 256)               1024      
_________________________________________________________________
dense_6 (Dense)              (None, 128)               32896     
_________________________________________________________________
dense_7 (Dense)              (None, 1)                 129       
Total params: 393,217
Trainable params: 391,681
Non-trainable params: 1,536
____________________________________________

In [0]:
# model.fit(x, y, callbacks=[roc_callback(training_data=(x, y)), tensorboard_callback], epochs=15)              

In [40]:
model.fit(X_tr, y_tr, validation_data=(X_val, y_val), callbacks=[roc_callback(training_data=(X_tr, y_tr),validation_data=(X_val, y_val))], epochs=5)

Train on 395661 samples, validate on 194879 samples
Epoch 1/5
roc-auc: 0.8292 - roc-auc_val: 0.8251                                                                                                    
roc-auc: 0.8292 - roc-auc_val: 0.8251                                                                                                    
395661/395661 [==============================] - 110s 279us/sample - loss: 0.1105 - acc: 0.9716 - val_loss: 0.1103 - val_acc: 0.9724
Epoch 2/5
roc-auc: 0.841 - roc-auc_val: 0.838                                                                                                    
roc-auc: 0.841 - roc-auc_val: 0.838                                                                                                    
395661/395661 [==============================] - 110s 278us/sample - loss: 0.1089 - acc: 0.9724 - val_loss: 0.1082 - val_acc: 0.9726
Epoch 3/5
roc-auc: 0.8399 - roc-auc_val: 0.8413                                                                   

- Find problem of value error. So increase columns and feel faulty column
- Problem with 442 columns but not with 440
- Probel is in 441:442
- For now I just use data till 440

In [0]:
%load_ext tensorboard
%tensorboard --logdir logs